File: Arrays/Rotate_An_Array/Python/rotateAnArray.py
import sys
import os
sys.path.append(os.path.dirname(os.path.dirname(sys.path[0]))) #Adding the root directory (topic directory) to sys path to make use of reverse array function

from Reverse_An_Array.python.reverseAnArray import ReverseAnArray


class RotateAnArray(ReverseAnArray):

    num_rotations = 0
    choice = ''
    a_end = None
    b_start = None

    def get_user_input(self):
        super().get_user_input()
        self.num_rotations = int(input('Enter the number of rotations you want to make: '))

    def calc_subset_based_on_direction(self):        
        if input('Left or Right ? (L/R): ').upper() == 'L':
            self.a_end = self.num_rotations -1
            self.b_start = self.num_rotations
        else:
            self.a_end = len(self.arr) - self.num_rotations - 1
            self.b_start = len(self.arr) - self.num_rotations


    def rotate_array(self):
        super().reverse_an_array(0, self.a_end)
        super().reverse_an_array(self.b_start, len(self.arr)-1)
        super().reverse_an_array()
        

File: Arrays/Rotate_An_Array/C++/RotateAnArray.cpp
#include <iostream>
#include "../../Reverse_An_Array/C++/ReverseAnArray.cpp"

using namespace std;

class RotateAnArray : public ReverseAnArray {

    protected:
        int num_rotations;
        char choice;
        int a_end;
        int b_start;

    public:
        void getArrayRotationInputs(){
            cout << "Enter the number of rotations: ";
            cin >> num_rotations;
            cout << "Left/Right? (L/R): ";
            cin >> choice;
        }

    public:
        void calc_subsets_based_on_directions(){
            if ( choice == 'L' ) {
                a_end = num_rotations - 1;                
                b_start = num_rotations;
            } else if ( choice == 'R' ) {
                a_end = n - num_rotations - 1;
                b_start = n - num_rotations;
            }
        }


    public:
        void rotateArray(){

            reverseArray(0, a_end);
            reverseArray(b_start, n-1);
            reverseArray();

        }

};